File: frontend/src/api/memory.ts
import client from './client';

export interface MemoryPool {
  id: string;
  tenant_id: string;
  name: string;
  description?: string;
  type: 'personal' | 'team' | 'system';
  purpose: 'conversation' | 'skill' | 'knowledge';
  priority: number; // 1-10
  max_tokens: number; // 0=不限
  auto_activate: boolean;
  trigger_rules?: string; // JSON
  owner_id?: string;
  enabled: boolean;
  memory_count: number;
  created_at: string;
  updated_at: string;
}

export interface MemoryEntry {
  id: string;
  pool_id: string;
  type: string;
  content: string;
  metadata?: string | null;
  sensitivity: string;
  created_at: string;
  updated_at: string;
}

export interface VectorMemory {
  id: string;
  tenant_id: string;
  pool_id: string;
  content: string;
  type: string;
  sensitivity: string;
  metadata?: Record<string, unknown>;
  created_at: string;
}

export interface UserMemory {
  id: string;
  tenant_id: string;
  user_id: string;
  pool_id?: string;
  type: string; // preference/fact/feedback
  content: string;
  content_hash?: string;
  source?: string;
  status?: string;
  entity_ids: string[];
  metadata: Record<string, unknown>;
  access_count: number;
  last_accessed_at: string | null;
  last_seen_at?: string | null;
  vector_indexed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface MemorySettings {
  id: string;
  tenant_id: string;
  user_id?: string | null;
  auto_extract_enabled: boolean;
  recall_enabled: boolean;
  sensitive_filter_enabled: boolean;
  audit_enabled: boolean;
  hybrid_search_enabled: boolean;
  hybrid_search_mode: string;
  created_at?: string;
  updated_at?: string;
}

export interface MemoryScoreBreakdown {
  memory_id: string;
  keyword_score: number;
  vector_score: number;
  recency_score: number;
  frequency_score: number;
  type_score: number;
  final_score: number;
  explanation: string;
}

export interface MemoryAuditLog {
  id: string;
  tenant_id: string;
  user_id: string;
  memory_id?: string | null;
  action: string;
  source: string;
  original_preview?: string;
  sanitized_preview?: string;
  reason?: string;
  metadata?: Record<string, unknown>;
  created_at: string;
}

export interface Entity {
  id: string;
  tenant_id: string;
  pool_id?: string;
  name: string;
  type: string; // tenant/user/connector/tool/skill
  ref_id: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface SkillMemory {
  id: string;
  tenant_id: string;
  user_id: string | null;
  pool_id?: string;
  name: string;
  description: string | null;
  content: string;
  category: string | null;
  tags: string[];
  usage_count: number;
  created_at: string;
  updated_at: string;
}

export interface MemoryStats {
  total_user_memories: number;
  total_session_memories: number;
  total_entities: number;
  total_skill_memories: number;
  by_type: Record<string, number>;
}

export const memoryApi = {
  // Memory pools
  listPools: (tenantId: string) =>
    client.get<MemoryPool[]>(`/tenants/${tenantId}/memory-pools`),
  createPool: (tenantId: string, data: Partial<MemoryPool>) =>
    client.post<MemoryPool>(`/tenants/${tenantId}/memory-pools`, data),
  getPool: (poolId: string) =>
    client.get<MemoryPool>(`/memory-pools/${poolId}`),
  listEntries: (poolId: string, limit = 50, type?: string) =>
    client.get<MemoryEntry[]>(`/memory-pools/${poolId}/entries`, { params: { limit, type } }),
  updatePool: (poolId: string, data: Partial<MemoryPool>) =>
    client.put<MemoryPool>(`/memory-pools/${poolId}`, data),
  deletePool: (poolId: string) =>
    client.delete(`/memory-pools/${poolId}`),

  // Vector memories
  listMemories: (tenantId: string, poolId?: string, limit = 50) =>
    client.get<{ memories: VectorMemory[] }>(`/tenants/${tenantId}/vector-memories`, { params: { pool_id: poolId, limit } }),
  saveMemory: (tenantId: string, data: { pool_id?: string; content: string; type?: string; sensitivity?: string }) =>
    client.post<VectorMemory>(`/tenants/${tenantId}/vector-memories`, data),
  searchMemories: (tenantId: string, query: string, poolId?: string, limit = 10) =>
    client.get<{ memories: VectorMemory[] }>(`/tenants/${tenantId}/vector-memories/search`, { params: { q: query, pool_id: poolId, limit } }),
  deleteMemory: (tenantId: string, memoryId: string) =>
    client.delete(`/tenants/${tenantId}/vector-memories/${memoryId}`),

  // User memories (all users in tenant)
  listAllUserMemories: (tenantId: string, limit = 50) =>
    client.get<{ memories: UserMemory[] }>(`/tenants/${tenantId}/user-memories`, { params: { limit } }),
  searchUserMemories: (tenantId: string, userId: string, query: string, limit = 10) =>
    client.get<{ memories: UserMemory[]; explanations: MemoryScoreBreakdown[] }>(`/tenants/${tenantId}/users/${userId}/memories/search`, { params: { q: query, limit } }),

  // Governance settings and audit logs
  getSettings: (tenantId: string, userId?: string) =>
    client.get<MemorySettings>(`/tenants/${tenantId}/memory-settings`, { params: { user_id: userId } }),
  updateSettings: (tenantId: string, data: Partial<MemorySettings>, userId?: string) =>
    client.put<MemorySettings>(`/tenants/${tenantId}/memory-settings`, data, { params: { user_id: userId } }),
  listAuditLogs: (tenantId: string, limit = 50) =>
    client.get<{ logs: MemoryAuditLog[] }>(`/tenants/${tenantId}/memory-audit-logs`, { params: { limit } }),

  // Entities
  listEntities: (tenantId: string, limit = 50) =>
    client.get<{ entities: Entity[] }>(`/tenants/${tenantId}/entities`, { params: { limit } }),

  // Skill memories
  listSkillMemories: (tenantId: string, limit = 50) =>
    client.get<{ memories: SkillMemory[] }>(`/tenants/${tenantId}/skill-memories`, { params: { limit } }),

  // Memory stats
  getStats: (tenantId: string) =>
    client.get<MemoryStats>(`/tenants/${tenantId}/memory-stats`),
};

File: internal/models/memory.go
package models

import (
	"encoding/json"
	"time"
)

// UserMemory 用户记忆
type UserMemory struct {
	ID             string          `json:"id" db:"id"`
	TenantID       string          `json:"tenant_id" db:"tenant_id"`
	UserID         string          `json:"user_id" db:"user_id"`
	PoolID         *string         `json:"pool_id" db:"pool_id"` // 所属记忆池
	Type           string          `json:"type" db:"type"` // preference/fact/feedback
	Content        string          `json:"content" db:"content"`
	Embedding      []byte          `json:"-" db:"embedding"`
	EntityIDs      json.RawMessage `json:"entity_ids" db:"entity_ids"`
	Metadata       json.RawMessage `json:"metadata" db:"metadata"`
	AccessCount    int             `json:"access_count" db:"access_count"`
	LastAccessedAt *time.Time      `json:"last_accessed_at" db:"last_accessed_at"`
	CreatedAt      time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at" db:"updated_at"`
}

// SessionMemory 会话记忆
type SessionMemory struct {
	ID        string          `json:"id" db:"id"`
	TenantID  string          `json:"tenant_id" db:"tenant_id"`
	UserID    string          `json:"user_id" db:"user_id"`
	SessionID string          `json:"session_id" db:"session_id"`
	Role      string          `json:"role" db:"role"` // user/assistant/system
	Content   string          `json:"content" db:"content"`
	Embedding []byte          `json:"-" db:"embedding"`
	TokenCount *int           `json:"token_count" db:"token_count"`
	EntityIDs json.RawMessage `json:"entity_ids" db:"entity_ids"`
	CreatedAt time.Time       `json:"created_at" db:"created_at"`
}

// Entity 实体
type Entity struct {
	ID        string          `json:"id" db:"id"`
	TenantID  string          `json:"tenant_id" db:"tenant_id"`
	PoolID    *string         `json:"pool_id" db:"pool_id"` // 所属记忆池
	Name      string          `json:"name" db:"name"`
	Type      string          `json:"type" db:"type"` // tenant/user/connector/tool/skill
	RefID     *string         `json:"ref_id" db:"ref_id"`
	Embedding []byte          `json:"-" db:"embedding"`
	Metadata  json.RawMessage `json:"metadata" db:"metadata"`
	CreatedAt time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt time.Time       `json:"updated_at" db:"updated_at"`
}

// EntityRelation 实体关系
type EntityRelation struct {
	ID             string          `json:"id" db:"id"`
	TenantID       string          `json:"tenant_id" db:"tenant_id"`
	SourceEntityID string          `json:"source_entity_id" db:"source_entity_id"`
	TargetEntityID string          `json:"target_entity_id" db:"target_entity_id"`
	RelationType   string          `json:"relation_type" db:"relation_type"` // belongs_to/uses/manages
	Metadata       json.RawMessage `json:"metadata" db:"metadata"`
	CreatedAt      time.Time       `json:"created_at" db:"created_at"`
}

// SkillMemory 技能记忆
type SkillMemory struct {
	ID          string          `json:"id" db:"id"`
	TenantID    string          `json:"tenant_id" db:"tenant_id"`
	UserID      *string         `json:"user_id" db:"user_id"`
	PoolID      *string         `json:"pool_id" db:"pool_id"` // 所属记忆池
	Name        string          `json:"name" db:"name"`
	Description *string         `json:"description" db:"description"`
	Content     string          `json:"content" db:"content"`
	Category    *string         `json:"category" db:"category"`
	Tags        json.RawMessage `json:"tags" db:"tags"`
	Embedding   []byte          `json:"-" db:"embedding"`
	UsageCount  int             `json:"usage_count" db:"usage_count"`
	CreatedAt   time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time       `json:"updated_at" db:"updated_at"`
}

// MemorySearchResult 记忆搜索结果
type MemorySearchResult struct {
	Memory    interface{} `json:"memory"`
	Score     float64     `json:"score"`
	Source    string      `json:"source"` // semantic/keyword/entity
}

// MemoryStats 记忆统计
type MemoryStats struct {
	TotalUserMemories    int `json:"total_user_memories"`
	TotalSessionMemories int `json:"total_session_memories"`
	TotalEntities        int `json:"total_entities"`
	TotalSkillMemories   int `json:"total_skill_memories"`
	ByType               map[string]int `json:"by_type"`
}
